packages/protocol/src/routers/XReceiveProxy.sol

```diff
@@ -0,0 +1,59 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity 0.8.15;
+
+/**
+ * @title XReceiveProxy
+ *
+ * @author Fujidao Labs
+ *
+ * @notice A proxy contract that receive the XReceive call from connext to
+ * allow assets to be safely pulled in {ConnextRouter.sol}
+ */
+
+import {SafeERC20} from "openzeppelin-contracts/contracts/token/ERC20/utils/SafeERC20.sol";
+import {IERC20} from "openzeppelin-contracts/contracts/token/ERC20/IERC20.sol";
+import {IXReceiver} from "../interfaces/connext/IConnext.sol";
+import {ConnextRouter} from "./ConnextRouter.sol";
+
+contract XReceiveProxy is IXReceiver {
+  using SafeERC20 for IERC20;
+
+  error XReceiveProxy__xReceive_noValueTransfer();
+  error XReceiveProxy__xReceive_notReceivedAssetBalance();
+
+  ConnextRouter public immutable connextRouter;
+
+  /**
+   * @notice Constructor that initializes XReceiveProxy
+   */
+  constructor(address connextRouter_) {
+    // No need to check address(0) since this contract is deployed within {ConnextRouter.sol}
+    connextRouter = ConnextRouter(payable(connextRouter_));
+  }
+
+  function xReceive(
+    bytes32 transferId,
+    uint256 amount,
+    address asset,
+    address originSender,
+    uint32 originDomain,
+    bytes memory callData
+  )
+    external
+    returns (bytes memory)
+  {
+    IERC20 asset_ = IERC20(asset);
+
+    if (amount == 0) {
+      revert XReceiveProxy__xReceive_noValueTransfer();
+    }
```

Comment:
Why passing calldata only (without transfer of funds) wouldn't be a valid case?

Reply:
You are right, Updating it.